tests/functionality/testLUSOL.cpp

```diff
@@ -0,0 +1,167 @@
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include <resolve/LinSolverDirectLUSOL.hpp>
+#include <resolve/matrix/Coo.hpp>
+#include <resolve/matrix/MatrixHandler.hpp>
+#include <resolve/matrix/io.hpp>
+#include <resolve/vector/Vector.hpp>
+#include <resolve/vector/VectorHandler.hpp>
+#include <resolve/workspace/LinAlgWorkspace.hpp>
+
+using namespace ReSolve::constants;
+using real_type = ReSolve::real_type;
+using vector_type = ReSolve::vector::Vector;
+
+real_type compute_error(int& error_sum,
+                        std::unique_ptr<ReSolve::MatrixHandler>& matrix_handler,
+                        std::unique_ptr<ReSolve::VectorHandler>& vector_handler,
+                        std::string matrix_path,
+                        std::string rhs_path)
+{
+  int status;
+
+  std::ifstream matrix_file(matrix_path);
+  if (!matrix_file.is_open()) {
+    std::cout << "Failed to open " << matrix_path << "\n";
+    return 1;
+  }
+
+  std::unique_ptr<ReSolve::matrix::Coo> A(ReSolve::io::readMatrixFromFile(matrix_file));
+  matrix_file.close();
+
+  A->expand();
+
+  std::ifstream rhs_file(rhs_path);
+  if (!rhs_file.is_open()) {
+    std::cout << "Failed to open " << rhs_path << "\n";
+    return 1;
+  }
+
+  std::unique_ptr<real_type[]> rhs(ReSolve::io::readRhsFromFile(rhs_file));
+  rhs_file.close();
+
+  ReSolve::LinSolverDirectLUSOL lusol;
+  std::unique_ptr<real_type[]> x(new real_type[A->getNumRows()]);
+  std::unique_ptr<vector_type> vec_rhs(new vector_type(A->getNumRows()));
+  std::unique_ptr<vector_type> vec_x(new vector_type(A->getNumRows()));
+  std::unique_ptr<vector_type> vec_r(new vector_type(A->getNumRows()));
+
+  vec_rhs->update(rhs.get(), ReSolve::memory::HOST, ReSolve::memory::HOST);
+  vec_rhs->setDataUpdated(ReSolve::memory::HOST);
+  vec_x->allocate(ReSolve::memory::HOST);
+
+  error_sum += lusol.setup(A.get());
+  error_sum += lusol.analyze();
+
+  status = lusol.factorize();
+  if (status != 0) {
+    // LUSOL will segfault if you continue to try to solve if factorization failed
+    error_sum += status;
+    return 1;
+  }
+
+  status = lusol.solve(vec_rhs.get(), vec_x.get());
+  if (status != 0) {
+    error_sum += status;
+    return 1;
+  }
+
+  std::unique_ptr<vector_type> vec_test(new vector_type(A->getNumRows()));
+  std::unique_ptr<vector_type> vec_diff(new vector_type(A->getNumRows()));
+
+  std::unique_ptr<real_type[]> x_data(new real_type[A->getNumRows()]);
+  std::fill_n(x_data.get(), A->getNumRows(), 1.0);
```

Comment:
I think it is better if we use raw pointers for raw data as we want to move it back and forth from GPUs. 
```suggestion
  real_type* x_data = new real_type[A->getNumRows()];
  std::fill_n(x_data, A->getNumRows(), 1.0);
```

Reply:
We need to make it explicit in the code that raw data should be accessed through raw pointers. This is a CPU-only test, so in this case it does not matter but as the best practice, we should not use unique pointers for the raw matrix and vector data.